tests/Unit/CrudPanel/CrudPanelColumnsTest.php

```diff
@@ -0,0 +1,207 @@
+        ],
+    ];
+
+    private $expectedTwoColumnsArray = [
+        'column1' => [
+            'name' => 'column1',
+            'label' => 'Column1',
+        ],
+        'column2' => [
+            'name' => 'column2',
+            'label' => 'Column2',
+        ],
+    ];
+
+    private $threeColumnsArray = [
+        [
+            'name' => 'column1',
+            'label' => 'Column1',
+        ],
+        [
+            'name' => 'column2',
+            'label' => 'Column2',
+        ],
+        [
+            'name' => 'column3',
+            'label' => 'Column3',
+        ],
+    ];
+
+    private $expectedThreeColumnsArray = [
+        'column1' => [
+            'name' => 'column1',
+            'label' => 'Column1',
+        ],
+        'column2' => [
+            'name' => 'column2',
+            'label' => 'Column2',
+        ],
+        'column3' => [
+            'name' => 'column3',
+            'label' => 'Column3',
+        ],
+    ];
+
+    public function testAddColumnByName()
+    {
+        $this->crudPanel->addColumn('column1');
+
+        $this->assertEquals(1, count($this->crudPanel->columns));
+        $this->assertContains($this->oneColumnArray, $this->crudPanel->columns);
+    }
+
+    public function testAddColumnsByName()
+    {
+        $this->crudPanel->addColumns(['column1', 'column2']);
+
+        $this->assertEquals(2, count($this->crudPanel->columns));
+        $this->assertEquals($this->expectedTwoColumnsArray, $this->crudPanel->columns);
+    }
+
+    public function testAddColumnAsArray()
+    {
+        $this->crudPanel->addColumn($this->oneColumnArray);
+
+        $this->assertEquals(1, count($this->crudPanel->columns));
+        $this->assertContains($this->oneColumnArray, $this->crudPanel->columns);
+    }
+
+    public function testAddColumnsAsArray()
+    {
+        $this->crudPanel->addColumns($this->twoColumnsArray);
+
+        $this->assertEquals(2, count($this->crudPanel->columns));
+        $this->assertEquals($this->expectedTwoColumnsArray, $this->crudPanel->columns);
+    }
+
+    public function testAddColumnNotArray()
+    {
+        $this->setExpectedException(\ErrorException::class);
+
+        $this->crudPanel->addColumns('column1');
+    }
+
+    public function testMoveColumnBefore()
+    {
+        $this->crudPanel->addColumns($this->twoColumnsArray);
+
+        $this->crudPanel->beforeColumn('column1');
+
+        $keys = array_keys($this->crudPanel->columns);
+        $this->assertEquals($this->expectedTwoColumnsArray['column2'], $this->crudPanel->columns[$keys[0]]);
+        $this->assertEquals(['column2', 'column1'], $keys);
+    }
+
+    public function testMoveColumnBeforeUnknownColumnName()
+    {
+        $this->markTestIncomplete('Not correctly implemented');
+
+        $this->crudPanel->addColumns($this->twoColumnsArray);
+
+        // TODO: fix the before column. it should not change the columns if the given target column does not exist.
```

Comment:
It does? Is there a bug / issue number for this?

Reply:
Not at the moment. I've discussed with @tabacitu and it's probably a good idea to just finish the unit tests and then take a look at all the things we've uncovered and create issues & fixes. 

Also, it's quite possible that some of this issues have already been fixed in some PRs, but I haven't got a chance to go through all of them yet.

Reply:
Yet, if there were a bug, I would have seen it, triaged it and possibly found the duplicates - in my experience TODOs with no bugs never get fixed and/or lost (trust me, there's so much to unit test that finishing could take months even with full-time effort).

@tabacitu - it really wouldn't take more than 3 minutes to copy paste the above TODO into a description, link to the pull request and maybe flag me directly to investigate?